proposals/3392/dataset_partitioning.md

```diff
@@ -0,0 +1,555 @@
+       },
+       "operator": "AND"
+     }
+   }
+}
+``` 
+
+An example of this would be a dataset partitioned by `business_date` and `country`. 
+A job affecting many countries for a single date would affect multiple 
+partitions and this would be described with a facet:
+
+```json
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "outputFacets": {
+      "subset": {
+         "$id": "https://openlineage.io/spec/facets/1-0-0/PartitionListSubsetCondition.json",
+         "partitions": [
+            {
+               "dimensions": {
+                  "business_date": "2024-10-15",
+                  "country": "PL"
+               }
+            },
+            {
+               "dimensions": {
+                  "business_date": "2024-10-15",
+                  "country": "DE"
+               }
+            }
+         ]
+      }
+   }
+}
+```
+
+A spark job:
+```python
+spark \
+  .read \
+  .parquet("/some/path/a=*/b=1") \
+  .filter(col("c") == 2)
+```
+
+would result in:
+
+```json
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "inputFacets": {
+      "subset": {
+         "type": "location",
+         "locations": [
+            "/some/path/a=1/b=1",
+            "/some/path/a=2/b=1",
+            "/some/path/a=3/b=1",
+            "/some/path/a=4/b=1"
+         ]
+      }
+   }
+}
+```
+
+If the number of locations exceeds a configurable threshold, the facet would be:
+```json
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "inputFacets": {
+      "subset": {
+         "type": "compare",
+         "left": {
+            "type": "field", "value": "c"
+         },
+         "right": {
+            "type": "literal", "value": "2"
+         },
+         "comparison": "EQUAL"
+      }
+   }
+}
+```
+
+## Reference implementation with Apache Spark
+
+ * Support extracting dataset partitioning info from:
+   * `DataSourceV2Relation objects`, see partitioning field ([see code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Table.java#L71)),
+   * [Iceberg partitioning](https://iceberg.apache.org/docs/1.6.1/partitioning/#problems-with-hive-partitioning).
+ * Verify how dataset filtering is reflected in `LogicalPlan`. Produce subset condition input facets based on filtering proposal prepared.
+ * Support configuration based partitioning:
+   * Provide configuration to manually specify how datasets are partitioned within the organization.
+   * For example, if a configuration contains `business_date` partitioning definition, then a Spark job reading data from `s3://bucket/dataset/business_date=20241101` would replace dataset identifier with `s3://bucket/dataset` and `FilterInputDatasetFacet` determining `business_date` specified.
+   * Similarly, a job writing a location `s3://bucket/dataset/business_date=20241101` would identify a dataset `s3://bucket/dataset` and enrich it with `PartitionOutputDatasetFacet`.
```

Comment:
Can we infer partitioning info from the path? For example, split location by `/`, all items without `=` are a part of dataset, and items with `=` are dataset subsets

Reply:
I think we can infer partitioning from the path. One can allow detecting partition dimensions from substrings `/a=b/`, but it may be also useful to limit this to datasets matching a specified regular expression. One may also disable auto partitioning detection for specific `a` values, or enable only for selected ones.  

It can become complex to describe all the partitioning strategies within the organisation, but looks doable.

Reply:
I'm not very optimistic about `describe all the partitioning strategies within the organisation`, but this is about Spark integration implementation detail, and not about facet proposal, so we can move this discussion elsewhere